--- TI-DSPLib/include/DSPLib_support.h.orig	2020-04-16 13:21:22.371001818 +0800
+++ TI-DSPLib/include/DSPLib_support.h	2020-04-16 13:22:08.726182783 +0800
@@ -328,6 +328,8 @@
 #endif //MSP_DISABLE_DIAGNOSTICS
 }
 
+#if defined(MSP_USE_LEA)
+
 // Benchmark register support
 #define HWREG16(x)          (*((volatile uint16_t *)((uint16_t)x)))
 #define BENCHMARK_TAxR      (0x0010)
@@ -395,6 +397,8 @@
     return (scale*((uint32_t)cycles << shift));
 }
 
+#endif
+
 #if defined(__MSP430_HAS_MPY32__)
 
 //******************************************************************************
