diff -ur ../../github/DSPLib_1_30_00_02/include/DSPLib_support.h TI-DSPLib/include/DSPLib_support.h
--- ../../github/DSPLib_1_30_00_02/include/DSPLib_support.h	2018-05-02 08:11:56.000000000 +0800
+++ TI-DSPLib/include/DSPLib_support.h	2020-07-24 11:52:01.925789743 +0800
@@ -252,8 +252,14 @@
 #if defined(__TI_COMPILER_VERSION__)
 #define _PRAGMA(x) _Pragma (#x)
 #if defined(MSP_USE_LEA)
+// https://processors.wiki.ti.com/index.php/Pragmas_in_C++
+#ifdef __cplusplus
+#define DSPLIB_DATA(var,align)  _PRAGMA(DATA_SECTION(".leaRAM"))\
+                                _PRAGMA(DATA_ALIGN(var,(align)))
+#else
 #define DSPLIB_DATA(var,align)  _PRAGMA(DATA_SECTION(var,".leaRAM"))\
                                 _PRAGMA(DATA_ALIGN(var,(align)))
+#endif
 #else
 #define DSPLIB_DATA(var,align)  _PRAGMA(DATA_ALIGN(var,(align)))
 #endif
